File: 73-set-matrix-zeroes/set-matrix-zeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int numRows=matrix.size();
        int numCols=matrix[0].size();
        vector<int> zeroCols(numCols,0);
        vector<int> zeroRows(numRows,0);

        for(int i=0;i<numRows;i++){
            for(int j=0;j<numCols;j++){
                if(matrix[i][j]==0){
                    zeroRows[i]=1;
                    zeroCols[j]=1;

                }
            }
        }

        for(int i=0;i<numRows;i++){
            for(int j=0;j<numCols;j++){
                if(zeroRows[i] ||zeroCols[j]){
                  matrix[i][j]=0;

                }
            }
        }
        
    }
};

File: 73-set-matrix-zeroes/set-matrix-zeroes.rs
impl Solution {
    pub fn set_zeroes(matrix: &mut Vec<Vec<i32>>) {
        if matrix.is_empty() || matrix[0].is_empty() {
            return;
        }

        let mut first_row_zero = false;
        let mut first_col_zero = false;

        let num_rows = matrix.len();
        let num_cols = matrix[0].len();

        for j in 0..num_cols {
            if matrix[0][j] == 0 {
                first_row_zero = true;
                break;
            }
        }

        for i in 0..num_rows {
            if matrix[i][0] == 0 {
                first_col_zero = true;
                break;
            }
        }

        for i in 1..num_rows {
            for j in 1..num_cols {
                if matrix[i][j] == 0 {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }

        for i in 1..num_rows {
            for j in 1..num_cols {
                if matrix[i][0] == 0 || matrix[0][j] == 0 {
                    matrix[i][j] = 0;
                }
            }
        }

        if first_row_zero {
            for j in 0..num_cols {
                matrix[0][j] = 0;
            }
        }

        if first_col_zero {
            for i in 0..num_rows {
                matrix[i][0] = 0;
            }
        }
    }
}